pandas/_typing.py

```diff
@@ -52,7 +52,7 @@
 # `def func(a: FrameOrSeriesUnion) -> FrameOrSeriesUnion: ...` means that if a Series
 # is passed in, either a Series or DataFrame is returned, and if a DataFrame is passed
 # in, either a DataFrame or a Series is returned.
-FrameOrSeriesUnion = Union["DataFrame", "Series"]
+FrameOrSeriesUnion = Union["DataFrame", "NDFrame", "Series"]
```

Comment:
I _think_ ommiting `NDFrame` from `FrameOrSeriesUnion` is intentional. can you revert this and replace `FrameOrSeriesUnion`  with `NDFrame` in the code where using `FrameOrSeriesUnion` is giving mypy errors.

Reply:
hi @simonjayhawkins, thanks a lot for review.

I'm struggling a bit to get mypy to work here. The problem I'm having is not with my own type hints, but the following line:
```python
diff = concat([self, other], axis=axis, keys=keys)
```
This gives me an error because `concat` accepts `FrameOrSeriesUnion`, but `self` has to be `NDFrame` or its parent class.
```python
def concat(
    objs: Union[Iterable[FrameOrSeriesUnion], Mapping[Label, FrameOrSeriesUnion]],
    ...
)
```

Do you have any suggestions how I should type hint `self` and `other` in my implementation in generic.py?

Reply:
yeah, there looks to be an issue with concat type annotations. It's only the return type that should be `FrameOrSeriesUnion`, the arguments should be `FrameOrSeries` which includes `NDFrame`. this patch should fix.

```diff
diff --git a/pandas/_typing.py b/pandas/_typing.py
index 7387e100b..3b7392f78 100644
--- a/pandas/_typing.py
+++ b/pandas/_typing.py
@@ -52,7 +52,7 @@ FilePathOrBuffer = Union[str, Path, IO[AnyStr]]
 # `def func(a: FrameOrSeriesUnion) -> FrameOrSeriesUnion: ...` means that if a Series
 # is passed in, either a Series or DataFrame is returned, and if a DataFrame is passed
 # in, either a DataFrame or a Series is returned.
-FrameOrSeriesUnion = Union["DataFrame", "NDFrame", "Series"]
+FrameOrSeriesUnion = Union["DataFrame", "Series"]
 
 # FrameOrSeries is stricter and ensures that the same subclass of NDFrame always is
 # used. E.g. `def func(a: FrameOrSeries) -> FrameOrSeries: ...` means that if a
diff --git a/pandas/core/reshape/concat.py b/pandas/core/reshape/concat.py
index 091129707..eb195f836 100644
--- a/pandas/core/reshape/concat.py
+++ b/pandas/core/reshape/concat.py
@@ -6,7 +6,7 @@ from typing import Iterable, List, Mapping, Union, overload
 
 import numpy as np
 
-from pandas._typing import FrameOrSeriesUnion, Label
+from pandas._typing import FrameOrSeries, FrameOrSeriesUnion, Label
 
 from pandas.core.dtypes.generic import ABCDataFrame, ABCSeries
 
@@ -48,7 +48,7 @@ def concat(
 
 @overload
 def concat(
-    objs: Union[Iterable[FrameOrSeriesUnion], Mapping[Label, FrameOrSeriesUnion]],
+    objs: Union[Iterable[FrameOrSeries], Mapping[Label, FrameOrSeries]],
     axis=0,
     join: str = "outer",
     ignore_index: bool = False,
@@ -63,7 +63,7 @@ def concat(
 
 
 def concat(
-    objs: Union[Iterable[FrameOrSeriesUnion], Mapping[Label, FrameOrSeriesUnion]],
+    objs: Union[Iterable[FrameOrSeries], Mapping[Label, FrameOrSeries]],
     axis=0,
     join="outer",
     ignore_index: bool = False,
```

Reply:
Applied the patch and all mypy issues resolved.
Thanks a lot for this!